feat(cli): implements 'safe files tree' subcommand

dan-da opened the pull request:
This PR addresses #409. 

I added `safe files tree` subcommand, which displays output in same style as unix `tree` command, but without any color highlighting.  Serialized output is also available.

I included a flag `--flat` that causes the full tree to be displayed in a flat list, with full paths displayed as well as file details.  This was very easy to do because that is the format returned by files_container_get() API, so I just exposed it without any filtering.

Example output of `safe files tree`

```
$ safe files tree safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbncsafe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc
├── another.md
├── noextension
├── subfolder
│   ├── sub2.md
│   └── subexists.md
└── test.md

$ safe files tree -o yaml safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc
---
name: "safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc"
sub:
  - name: another.md
  - name: noextension
  - name: subfolder
    sub:
      - name: sub2.md
      - name: subexists.md
  - name: test.md

$ safe files tree -o json safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc
{
  "name": "safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc",
  "sub": [
    {
      "name": "another.md"
    },
    {
      "name": "noextension"
    },
    {
      "name": "subfolder",
      "sub": [
        {
          "name": "sub2.md"
        },
        {
          "name": "subexists.md"
        }
      ]
    },
    {
      "name": "test.md"
    }
  ]
}
```

Example output of `safe files tree --flat`

```
$ safe files tree --flat safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc
Files of FilesContainer (version 0) at "safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc":
Total: 45
SIZE  CREATED               MODIFIED              NAME 
6     2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  /another.md 
0     2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  /noextension 
4     2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  /subfolder/sub2.md 
23    2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  /subfolder/subexists.md 
12    2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  /test.md

$ safe files tree --flat --json safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc 
{
  "/another.md": {
    "created": "2020-03-03T22:31:21Z",
    "link": "safe://hbhyrydt5b95dmumcm8yig4u1keuuh8hgsr5yx39xn4mqikp91sbdhbpwp",
    "modified": "2020-03-03T22:31:21Z",
    "size": "6",
    "type": "text/x-markdown"
  },
  "/noextension": {
    "created": "2020-03-03T22:31:21Z",
    "link": "safe://hbyyyynw7fi7ko8u3wtai6eekko3m5y6u9ud7w1i4ajycmwfwu3wwpnn81",
    "modified": "2020-03-03T22:31:21Z",
    "size": "0",
    "type": "Raw"
  },
  "/subfolder/sub2.md": {
    "created": "2020-03-03T22:31:21Z",
    "link": "safe://hbhyrydyr41tz98funce1eqzkgju8e1yb4343ppct4fiwie5n3qhsw7o7x",
    "modified": "2020-03-03T22:31:21Z",
    "size": "4",
    "type": "text/x-markdown"
  },
  "/subfolder/subexists.md": {
    "created": "2020-03-03T22:31:21Z",
    "link": "safe://hbhyryn9uodh1ju5uzyti3gmmtwburrssd89rcwcy3rzofdpypwomrzzte",
    "modified": "2020-03-03T22:31:21Z",
    "size": "23",
    "type": "text/x-markdown"
  },
  "/test.md": {
    "created": "2020-03-03T22:31:21Z",
    "link": "safe://hbhyrydpan7d94mwp1bun3mxfnrfrui131an7ihu11wsn8dkr8odab9qwn",
    "modified": "2020-03-03T22:31:21Z",
    "size": "12",
    "type": "text/x-markdown"
  }
}
```

**Testing**

I created a test case which tests output of `safe files tree` and `safe files tree --json` using the testdata directory.

**Notes**
* When using `--flat`, the paths are prefixed with '/'.  files_container_get() API returns them this way, and I think perhaps it should be modified to not do that.  The alternative is to copy/modify the list, which can be slow for very large trees.
* empty directories are not included in the list returned from files_container_get() even though one exists in the tree that was originally `safe files put`.  I will file an issue about this.
* The name of the root folder is displayed as a SAFE xorurl.  Yet it had a name when it was `safe files put`. So it seems we have lost information/structure.  If I use zip/unzip, tar/untar, cp -r, rsync, etc, I am able to preserve the root dirname.   I see no reason why `safe` should be different.  I will file an issue for this also.
* The implementation uses recursion to generate the tree.  This could blow up with a deep enough tree.  An iterative approach using a heap stack would be better.  I tried this initially and couldn't get it to work with the borrow checker.  Perhaps now that the recursive version works, it might be easier.  Anyway, I doubt the recursive version will run into any problems in the near future as people don't normally have directory structures thousands of levels deep.
* see additional notes/comments in the code.

bochaco commented on safe-cli/subcommands/files.rs:
```diff
@@ -311,6 +388,112 @@ fn print_serialized_output(
+                };
+                let n = FileSystemNode::new(&item, fs_type);
+                node.add_child(n);
+                // Very gross, but it works.
```
I think you just need to change the above line to `node.add_child(n.clone());` (you may need to add `Clone` to the derive clause of the struct), then you should be able to `n`

dan-da replied:
Well, but this function is building a single tree recursively.  We need to reference the tree node, not copy/clone it.

bochaco commented:
Also note, the total number of files you output for the `--flat` example seems wrong, it says 45 but you have only 5 files in total.

I'm thinking it could be nice though (not suggesting part of the scope of this PR) to have a flag which we get the files info with the tree format, e.g.:
```
Files of FilesContainer (version 0) at "safe://hnyynysurwzoiudugs47d7fjx7hwg3i7ch1ryowdm7gpgj4538n1mwqukgbnc"
Total: 5
SIZE  CREATED               MODIFIED              NAME 
                                                  /
6     2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  ├── another.md
0     2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  ├── noextension 
                                                  ├── subfolder
4     2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  │   ├── sub2.md
23    2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  │   └── subexists.md
12    2020-03-03T22:31:21Z  2020-03-03T22:31:21Z  └── test.md
```

bochaco commented:
@dan-da, I was thinking this could be better/simpler if you define the nodes as:
```
type SubNodes = Vec<Box<FilesTreeNode>>;

enum FilesTreeNode {
    File(String),
    Directory((String, SubNodes)),
}
```
It may make things simpler I believe.

dan-da commented:
I've updated this PR.  Changes:
1. rebased to latest master to fix merge conflicts.
2. added `--details` flag to `files tree`
3. removed `--flat` flag.  I will turn it into `safe files ls --recursive` in a separate PR.
4. add additional Totals columns for `safe files ls`

*Sample output:*

```
$ safe files tree --details  safe://hnyynysebzh45brmgkpbfh4tthejr5qkxa57nh8dg61djc4g4ec61xpdoybnc
SIZE  CREATED               MODIFIED              NAME 
                                                  safe://hnyynysebzh45brmgkpbfh4tthejr5qkxa57nh8dg61djc4g4ec61xpdoybnc 
6     2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  ├── another.md 
0     2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  ├── noextension 
                                                  ├── subfolder 
4     2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  │   ├── sub2.md 
23    2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  │   └── subexists.md 
12    2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  └── test.md
```

```
$ safe files tree --details -o yaml  safe://hnyynysebzh45brmgkpbfh4tthejr5qkxa57nh8dg61djc4g4ec61xpdoybnc
---
name: "safe://hnyynysebzh45brmgkpbfh4tthejr5qkxa57nh8dg61djc4g4ec61xpdoybnc"
sub:
  - name: another.md
    details:
      created: "2020-03-05T19:49:03Z"
      link: "safe://hbhynydt5b95dmumcm8yig4u1keuuh8hgsr5yx39xn4mqikp91sbdhbpwp"
      modified: "2020-03-05T19:49:03Z"
      size: "6"
      type: text/markdown
  - name: noextension
    details:
      created: "2020-03-05T19:49:03Z"
      link: "safe://hbyyyynw7fi7ko8u3wtai6eekko3m5y6u9ud7w1i4ajycmwfwu3wwpnn81"
      modified: "2020-03-05T19:49:03Z"
      size: "0"
      type: Raw
  - name: subfolder
    sub:
      - name: sub2.md
        details:
          created: "2020-03-05T19:49:03Z"
          link: "safe://hbhynydyr41tz98funce1eqzkgju8e1yb4343ppct4fiwie5n3qhsw7o7x"
          modified: "2020-03-05T19:49:03Z"
          size: "4"
          type: text/markdown
      - name: subexists.md
        details:
          created: "2020-03-05T19:49:03Z"
          link: "safe://hbhynyn9uodh1ju5uzyti3gmmtwburrssd89rcwcy3rzofdpypwomrzzte"
          modified: "2020-03-05T19:49:03Z"
          size: "23"
          type: text/markdown
  - name: test.md
    details:
      created: "2020-03-05T19:49:03Z"
      link: "safe://hbhynydpan7d94mwp1bun3mxfnrfrui131an7ihu11wsn8dkr8odab9qwn"
      modified: "2020-03-05T19:49:03Z"
      size: "12"
      type: text/markdown
```
```
$ safe files ls  safe://hnyynysebzh45brmgkpbfh4tthejr5qkxa57nh8dg61djc4g4ec61xpdoybncFiles of FilesContainer (version 0) at "safe://hnyynysebzh45brmgkpbfh4tthejr5qkxa57nh8dg61djc4g4ec61xpdoybnc":
Files: 3   Size: 18   Total Files: 5   Total Size: 45
SIZE  CREATED               MODIFIED              NAME 
12    2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  est.md 
6     2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  nother.md 
0     2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  oextension 
27    2020-03-05T19:49:03Z  2020-03-05T19:49:03Z  ubfolder/
```

dan-da commented on safe-cli/subcommands/files.rs:
```diff
@@ -16,9 +16,69 @@ use super::{
+
+impl FileSystemNode {
+    // create a new FileSystemNode (either a Directory or File)
+    fn new(name: &str, fs_type: FileSystemNodeType) -> FileSystemNode {
```
changed.

dan-da commented:
Updated PR again.
* Added color highlight (blue) for directories (only active if stdout is a TTY)
* Added a summary row with directory, file counts, to match unix tree output.
* squashed all commits into one.

dan-da commented on safe-cli/subcommands/files.rs:
```diff
@@ -16,9 +16,69 @@ use super::{
+// A recursive type to represent a directory tree.
+// used by `safe files tree`
+#[derive(Debug, Serialize)]
+struct FileSystemNode {
```
I changed it to FilesTreeNode

dan-da commented:
Updated again:
* changed FileSystemNode to FileTreeNode
* added result filtering by xorurl path.

comment:  I believe that result filtering by url path is such an expected behavior that it should either be made the default behavior for file_container_get() or add a new api file_container_get_matches() (maybe cleaner) that does it.  I can do this in separate PR if concept approved.
